app/services/exchange_deleter.rb

```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+class ExchangeDeleter
+  def delete(child, parent)
+    main_exchanges = Exchange.
+      joins(:order_cycle).
+      where(sender: parent, receiver: child)
+    main_exchanges.destroy_all
```

Comment:
Do you still need the variable `main_exchanges`? Maybe you could call the method `delete_incoming`. And would it be useful to define `incoming: true` in the query? I can't think of a scenario where an outgoing or internal exchange would have these parent and receiver values but it could be another safeguard and even speed up the query with an indexed boolean variable.